File: internal/wasmhost/pool.go
package wasmhost

import (
	"container/list"
	"context"
	"sync"
	"time"

	"github.com/tetratelabs/wazero/api"
)

// instanceKey is what an idle instance can be handed back out for.
//
// The principal is in the key deliberately, and it is a departure from "LRU per
// (app, version)". A warm instance IS a cache of guest memory: whatever the
// previous call left in the guest's globals, heap or scratch buffers is still
// sitting there. Handing that instance to a different principal is an isolation
// break of exactly the kind D17.7 rejected one layer up, where the memo cache
// gained a principal in its key for the same reason. Cross-principal warmth is
// not worth it. Two people using the same assistant identity do not share
// memory (D13.4), and that has to be true of the runtime too, not just of rows.
//
// The cost is a colder pool per principal. At family scale the principal count
// is single digits, so the pool fragments by a small constant.
type instanceKey struct {
	moduleHash string
	principal  PrincipalID
	// tier, because an instance belongs to the runtime that made it. Two
	// runtimes differ in memory ceiling and in whether termination checks are
	// compiled in, so their instances are not interchangeable.
	tier tierKey
}

type instance struct {
	key      instanceKey
	mod      api.Module
	app      string
	lastUsed time.Time

	// bytes is the instance's wasm memory at the moment it went idle. Memory
	// grows during a call and never shrinks, so this is recomputed on release
	// rather than measured once at instantiation.
	bytes uint64

	// dead marks an instance that must never be reused: it trapped, it was
	// terminated, or its module was closed. A terminated instance is dead, not
	// paused.
	dead bool

	elem *list.Element
}

// instanceOverheadBytes is what wazero allocates per instance beyond the
// guest's linear memory: tables, globals, the module instance itself.
//
// Measured at ~10 KB against 128 KB of linear memory (BenchmarkInstanceFootprint,
// docs/wasmhost-benchmarks.md), which confirms the premise that memory
// dominates and a count-bounded pool would measure the wrong thing. It is
// counted anyway so a 512MB budget means 512MB rather than 512MB plus eight
// percent.
const instanceOverheadBytes = 16 << 10

// memBytes reports the instance's current wasm memory size.
func (i *instance) memBytes() uint64 {
	if i.mod == nil {
		return 0
	}
	mem := i.mod.Memory()
	if mem == nil {
		return instanceOverheadBytes
	}
	return uint64(mem.Size()) + instanceOverheadBytes
}

func (i *instance) close(ctx context.Context) {
	i.dead = true
	if i.mod != nil {
		_ = i.mod.Close(ctx)
	}
}

// pool is an LRU of idle instances bounded by SUMMED wasm memory rather than by
// instance count, because memory dominates the per-instance footprint and a
// count-bounded pool measures the wrong thing.
//
// Eviction and kill are the same primitive: Module.Close.
type pool struct {
	mu     sync.Mutex
	idle   map[instanceKey][]*instance
	lru    *list.List // front = most recently used
	bytes  uint64
	budget uint64
	ttl    time.Duration
	closed bool
}

func newPool(budget uint64, ttl time.Duration) *pool {
	return &pool{
		idle:   make(map[instanceKey][]*instance),
		lru:    list.New(),
		budget: budget,
		ttl:    ttl,
	}
}

// acquire takes a warm instance out of the pool, or returns nil if there is
// none. The caller instantiates on nil; the pool never instantiates, because it
// has no runtime and should not grow one.
func (p *pool) acquire(key instanceKey) *instance {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.closed {
		return nil
	}
	stack := p.idle[key]
	if len(stack) == 0 {
		return nil
	}
	// Take from the end: the hottest instance for this key, so a burst of calls
	// reuses one instance rather than touching every idle one in turn.
	inst := stack[len(stack)-1]
	p.idle[key] = stack[:len(stack)-1]
	if len(p.idle[key]) == 0 {
		delete(p.idle, key)
	}
	p.remove(inst)
	return inst
}

// release returns an instance to the pool, or closes it if it is dead or the
// pool is full. Safe to call on nil.
func (p *pool) release(ctx context.Context, inst *instance) {
	if inst == nil {
		return
	}
	if inst.dead {
		inst.close(ctx)
		return
	}

	p.mu.Lock()
	if p.closed {
		p.mu.Unlock()
		inst.close(ctx)
		return
	}
	inst.bytes = inst.memBytes()
	inst.lastUsed = time.Now()
	p.idle[inst.key] = append(p.idle[inst.key], inst)
	inst.elem = p.lru.PushFront(inst)
	p.bytes += inst.bytes

	evicted := p.evictLocked()
	p.mu.Unlock()

	for _, e := range evicted {
		e.close(ctx)
	}
}

// evictLocked drops least-recently-used instances until the summed memory of
// the idle set is back inside budget. Callers close the returned instances
// outside the lock: Module.Close can block, and holding the pool lock across it
// would stall every other call in the process.
func (p *pool) evictLocked() []*instance {
	var evicted []*instance
	for p.bytes > p.budget {
		back := p.lru.Back()
		if back == nil {
			break
		}
		inst, ok := back.Value.(*instance)
		if !ok {
			p.lru.Remove(back)
			continue
		}
		p.detachLocked(inst)
		evicted = append(evicted, inst)
	}
	return evicted
}

// sweep closes instances that have sat idle past the TTL.
func (p *pool) sweep(ctx context.Context) {
	cutoff := time.Now().Add(-p.ttl)
	p.mu.Lock()
	if p.closed {
		p.mu.Unlock()
		return
	}
	var expired []*instance
	for e := p.lru.Back(); e != nil; {
		prev := e.Prev()
		inst, ok := e.Value.(*instance)
		if !ok || inst.lastUsed.After(cutoff) {
			// The list is ordered by use, so the first fresh one ends the sweep.
			break
		}
		p.detachLocked(inst)
		expired = append(expired, inst)
		e = prev
	}
	p.mu.Unlock()

	for _, inst := range expired {
		inst.close(ctx)
	}
}

func (p *pool) closeAll(ctx context.Context) {
	p.mu.Lock()
	p.closed = true
	all := make([]*instance, 0, p.lru.Len())
	for e := p.lru.Front(); e != nil; e = e.Next() {
		if inst, ok := e.Value.(*instance); ok {
			all = append(all, inst)
		}
	}
	p.idle = make(map[instanceKey][]*instance)
	p.lru.Init()
	p.bytes = 0
	p.mu.Unlock()

	for _, inst := range all {
		inst.close(ctx)
	}
}

func (p *pool) stats() (int, uint64) {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.lru.Len(), p.bytes
}

// remove takes an instance out of the LRU list and the byte total, leaving the
// idle map to the caller.
func (p *pool) remove(inst *instance) {
	if inst.elem != nil {
		p.lru.Remove(inst.elem)
		inst.elem = nil
	}
	p.bytes -= inst.bytes
}

// detachLocked takes an instance out of both the LRU list and the idle map.
func (p *pool) detachLocked(inst *instance) {
	p.remove(inst)
	stack := p.idle[inst.key]
	for i, cand := range stack {
		if cand == inst {
			p.idle[inst.key] = append(stack[:i], stack[i+1:]...)
			break
		}
	}
	if len(p.idle[inst.key]) == 0 {
		delete(p.idle, inst.key)
	}
}

File: crates/hive-wasmhost/src/pool.rs
//! The idle-instance pool: an LRU bounded by SUMMED wasm memory rather than by
//! instance count, because memory dominates the per-instance footprint.

use std::collections::HashMap;
use std::time::{Duration, Instant};

use parking_lot::Mutex;
use uuid::Uuid;
use wasmtime::{Memory, Store};

use crate::host::{State, TierKey};
use crate::limiter::LimiterError;

/// What an idle instance can be handed back out for.
///
/// The principal is in the key deliberately. A warm instance IS a cache of
/// guest memory, and handing one principal's leftover heap to another is an
/// isolation break (D17.7 one layer up). The capability bits are in it because
/// the link check used to run only on a pool miss, and a revoked grant kept
/// working for as long as the instance stayed warm.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub(crate) struct InstanceKey {
    pub(crate) module_hash: String,
    pub(crate) principal: Uuid,
    pub(crate) tier: TierKey,
    pub(crate) caps: u32,
}

/// What wasmtime allocates per instance beyond the guest's linear memory.
/// Deliberately above the measured range: over-counting holds slightly fewer
/// instances, under-counting quietly overcommits the box.
pub(crate) const INSTANCE_OVERHEAD_BYTES: u64 = 32 << 10;

pub(crate) struct Instance {
    pub(crate) key: InstanceKey,
    pub(crate) store: Store<State>,
    pub(crate) instance: wasmtime::Instance,
    pub(crate) memory: Memory,
    pub(crate) last_used: Instant,
    /// Monotonic use counter: the LRU order without a linked list.
    seq: u64,
    /// The instance's memory at the moment it went idle. Memory grows during
    /// a call and never shrinks, so this is recomputed on release.
    bytes: u64,
    /// An instance that must never be reused: it trapped, it was terminated,
    /// or it handled untrusted bytes. Dead, not paused.
    pub(crate) dead: bool,
}

impl Instance {
    pub(crate) fn new(
        key: InstanceKey,
        store: Store<State>,
        instance: wasmtime::Instance,
        memory: Memory,
    ) -> Self {
        let mut inst = Instance {
            key,
            store,
            instance,
            memory,
            last_used: Instant::now(),
            seq: 0,
            bytes: 0,
            dead: false,
        };
        inst.bytes = inst.mem_bytes();
        inst
    }

    pub(crate) fn mem_bytes(&self) -> u64 {
        self.memory.data_size(&self.store) as u64 + INSTANCE_OVERHEAD_BYTES
    }
}

#[derive(Default)]
struct PoolInner {
    idle: HashMap<InstanceKey, Vec<Instance>>,
    count: usize,
    bytes: u64,
    reserved: u64,
    seq: u64,
    closed: bool,
}

/// Two budgets, not one (D9.3): reserved memory is subtracted from what the
/// evictable pool may hold, so pinning takes memory away from warmth,
/// visibly.
pub(crate) struct Pool {
    inner: Mutex<PoolInner>,
    budget: u64,
    reserved_budget: u64,
    ttl: Duration,
}

impl Pool {
    pub(crate) fn new(budget: u64, reserved_budget: u64, ttl: Duration) -> Pool {
        Pool {
            inner: Mutex::new(PoolInner::default()),
            budget,
            reserved_budget,
            ttl,
        }
    }

    fn evictable_budget(&self, p: &PoolInner) -> u64 {
        self.budget.saturating_sub(p.reserved)
    }

    /// Charges a pinned instance against the reserved budget, or refuses.
    pub(crate) fn reserve(&self, bytes: u64) -> Result<(), LimiterError> {
        let evicted = {
            let mut p = self.inner.lock();
            if p.closed {
                return Err(LimiterError::AtCapacity("pool closed".into()));
            }
            if self.reserved_budget > 0 && p.reserved + bytes > self.reserved_budget {
                return Err(LimiterError::AtCapacity(format!(
                    "pinned instances hold {} of {} reserved bytes, this one needs {bytes}",
                    p.reserved, self.reserved_budget
                )));
            }
            p.reserved += bytes;
            // Pinning shrinks what warmth may hold, so make room now.
            self.evict_locked(&mut p)
        };
        drop(evicted);
        Ok(())
    }

    pub(crate) fn unreserve(&self, bytes: u64) {
        let mut p = self.inner.lock();
        p.reserved = p.reserved.saturating_sub(bytes);
    }

    /// Takes the hottest warm instance for a key, or `None`. The pool never
    /// instantiates: it has no engine and should not grow one.
    pub(crate) fn acquire(&self, key: &InstanceKey) -> Option<Instance> {
        let mut p = self.inner.lock();
        if p.closed {
            return None;
        }
        let stack = p.idle.get_mut(key)?;
        let inst = stack.pop()?;
        if stack.is_empty() {
            p.idle.remove(key);
        }
        p.count -= 1;
        p.bytes -= inst.bytes;
        Some(inst)
    }

    /// Returns an instance to the pool, or drops it if it is dead or the pool
    /// is full. Dropping happens outside the lock: freeing a store can take a
    /// moment, and holding the lock across it would stall every other call.
    pub(crate) fn release(&self, mut inst: Instance) {
        if inst.dead {
            drop(inst);
            return;
        }
        let evicted = {
            let mut p = self.inner.lock();
            if p.closed {
                drop(p);
                drop(inst);
                return;
            }
            inst.bytes = inst.mem_bytes();
            inst.last_used = Instant::now();
            p.seq += 1;
            inst.seq = p.seq;
            p.count += 1;
            p.bytes += inst.bytes;
            p.idle.entry(inst.key.clone()).or_default().push(inst);
            self.evict_locked(&mut p)
        };
        drop(evicted);
    }

    /// Drops least-recently-used instances until the idle set is back inside
    /// budget. Returns them for the caller to drop outside the lock.
    fn evict_locked(&self, p: &mut PoolInner) -> Vec<Instance> {
        let mut out = Vec::new();
        while p.bytes > self.evictable_budget(p) {
            let Some(victim) = self.detach_oldest(p) else {
                break;
            };
            out.push(victim);
        }
        out
    }

    fn detach_oldest(&self, p: &mut PoolInner) -> Option<Instance> {
        let (key, idx) = p
            .idle
            .iter()
            .flat_map(|(k, v)| v.iter().enumerate().map(move |(i, inst)| (k, i, inst.seq)))
            .min_by_key(|(_, _, seq)| *seq)
            .map(|(k, i, _)| (k.clone(), i))?;
        let stack = p.idle.get_mut(&key)?;
        let inst = stack.remove(idx);
        if stack.is_empty() {
            p.idle.remove(&key);
        }
        p.count -= 1;
        p.bytes -= inst.bytes;
        Some(inst)
    }

    /// Drops instances that have sat idle past the TTL.
    pub(crate) fn sweep(&self) {
        let cutoff = Instant::now() - self.ttl;
        let expired = {
            let mut p = self.inner.lock();
            if p.closed {
                return;
            }
            let mut out = Vec::new();
            let mut freed_bytes = 0u64;
            for stack in p.idle.values_mut() {
                let mut i = 0;
                while i < stack.len() {
                    if stack[i].last_used <= cutoff {
                        let inst = stack.remove(i);
                        freed_bytes += inst.bytes;
                        out.push(inst);
                    } else {
                        i += 1;
                    }
                }
            }
            p.idle.retain(|_, v| !v.is_empty());
            p.count -= out.len();
            p.bytes -= freed_bytes;
            out
        };
        drop(expired);
    }

    pub(crate) fn close_all(&self) {
        let all: Vec<Instance> = {
            let mut p = self.inner.lock();
            p.closed = true;
            p.count = 0;
            p.bytes = 0;
            p.idle.drain().flat_map(|(_, v)| v).collect()
        };
        drop(all);
    }

    pub(crate) fn stats(&self) -> (usize, u64, u64) {
        let p = self.inner.lock();
        (p.count, p.bytes, p.reserved)
    }
}
